File: backend/app/api/categories.py
from typing import Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.db.engine import get_db
from app.schemas.category import CategoryCreate, CategoryListResponse, CategoryOut
from app.services.category import CategoryService

router = APIRouter()


@router.get("/categories", response_model=CategoryListResponse)
def list_categories(
    page: int = Query(1, ge=1),
    page_size: int = Query(100, ge=1, le=200),
    q: str | None = Query(None),
    sort_by: str = Query("name"),
    sort_dir: Literal["asc", "desc"] = Query("asc"),
    db: Session = Depends(get_db),
):
    service = CategoryService(db)
    return service.list_categories(
        page=page, page_size=page_size, q=q, sort_by=sort_by, sort_dir=sort_dir
    )


@router.post("/categories", response_model=CategoryOut, status_code=201)
def create_category(data: CategoryCreate, db: Session = Depends(get_db)):
    service = CategoryService(db)
    return service.create_category(data)

File: frontend/src/api/categories.ts
import { api } from "./client";

export interface Category {
  id: number;
  name: string;
  is_system: boolean;
  created_at: string;
}

export interface CategoryListResponse {
  items: Category[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListCategoriesParams {
  page?: number;
  page_size?: number;
  q?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
}

export function listCategories(params?: ListCategoriesParams): Promise<CategoryListResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.page_size) search.set("page_size", String(params.page_size));
  if (params?.q) search.set("q", params.q);
  if (params?.sort_by) search.set("sort_by", params.sort_by);
  if (params?.sort_dir) search.set("sort_dir", params.sort_dir);
  const qs = search.toString();
  return api.get<CategoryListResponse>(`/api/categories${qs ? `?${qs}` : ""}`);
}

export function createCategory(data: { name: string }): Promise<Category> {
  return api.post<Category>("/api/categories", data);
}
